src/autospawn/auto_spawn.ts

```diff
@@ -0,0 +1,76 @@
+
+export class AutoSpawn {
+    static CONSONANTS = <string>"bcdfghjklmnpqrstvwxz"
+
+    static VOWELS = <string>"aeiouy"
+
+    //Pick a random name
+    static generateName() {
+        let name = "";
+        for (let i = 0; i < 5; i++) {
+            if (i % 2 == 0) {
+                let letter = this.CONSONANTS[Math.floor(Math.random() * this.CONSONANTS.length)];
+                if (i == 0) {
+                    name += letter.toUpperCase();
+                } else {
+                    name += letter;
+                }
+            } else {
+                name += this.VOWELS[Math.floor(Math.random() * this.VOWELS.length)];
+            }
+        }
+        return name;
+    }
+
+    static addToSpawnQueue(room: Room, creepSpawnRequest: any) {
+        const roles = _.map(room.memory.spawnQueue, i => i.memory.role);
+        if (_.contains(roles, creepSpawnRequest.memory.role) == false) {
+            room.memory.spawnQueue.push(creepSpawnRequest);
+            console.log("Adding to spawn queue: " + JSON.stringify(creepSpawnRequest));
+        }
+    }
+
+    static queueSpawns(room: Room) {
+        room.memory.creepRoleCounts = room.memory.creepRoleCounts || { tick: 0, data: {} };
+        room.memory.spawnQueue = room.memory.spawnQueue || [];
+
+        //count each creep by role and store info
+        if (room.memory.creepRoleCounts.tick != Game.time) {
+            room.memory.creepRoleCounts.data = _.countBy(_.filter(Game.creeps, (c) => { return c.memory.spawnRoom == room.name; }), (c) => { return c.memory.role });
+            room.memory.creepRoleCounts.tick = Game.time;
+        }
+
+        //failsafe to restore room, spawn a single harvester to bootstrap the base
+        if (Object.keys(room.memory.creepRoleCounts.data).length == 0) {
+            const creepSpawnRequest = {
+                memory: <CreepMemory>{ role: 'harvester', task: null, working: false, spawnRoom: room.name },
+                body: [<BodyPartConstant>WORK, <BodyPartConstant>CARRY, <BodyPartConstant>MOVE]
```

Comment:
I think we can get rid of these types too, since typescript figures out the types. We just need to add correct types to the `creepSawnRequest` parameter in the `addToSpawnQueue` method.